docs/proposals/koordlet/20221018-koordlet-support-cgroups-v2.md

```diff
@@ -0,0 +1,461 @@
+So the `IsSupported()` method is defined for checking the compatibility of system resources when trying to modify the files.
+
+```go
+type ResourceType string
+
+type SystemResource interface {
+    // the type of system resource. e.g. "cpu/cpu.cfs_quota_us", "memory.min", "schemata"
+	ResourceType() ResourceType
+	// the generated system file path according to the given parent directory. e.g. "/host-cgroup/kubepods/kubepods-podxxx/cpu.shares"
+	Path(parentDir string) string
+	// whether the system resource is supported in current platform
+	IsSupported() bool
+	// whether the given value is valid for the system resource's content
+	IsValid(v string) (bool, string)
+	WithSupported(supported bool) SystemResource
+	WithValidator(validator ResourceValidator) SystemResource
+}
+
+type ResourceValidator interface {
+	Validate(value string) (isValid bool, msg string)
+}
+```
+
+The implementation for cgroup resources are shown below, which trim the old `CgroupFile`.
+
+```go
+type CgroupResource struct {
+	FileName  string
+	Subfs     string
+	Supported bool
+	Validator ResourceValidator
+}
+
+func (c *CgroupResource) ResourceType() ResourceType {
+	return GetCgroupResourceType(c.Subfs, c.FileName)
+}
+
+func (c *CgroupResource) Path(parentDir string) string {
+	return filepath.Join(sysutil.Conf.CgroupRootDir, c.Subfs, parentDir, c.FileName)
+}
+
+func (c *CgroupResource) IsSupported() bool {
+	return c.Supported
+}
+
+func (c *CgroupResource) IsValid(v string) (bool, string) {
+	if c.Validator == nil {
+		return true, ""
+	}
+	return c.Validator.Validate(v)
+}
+
+func (c *CgroupResource) WithSupported(supported bool) SystemResource {
+	c.Supported = supported
+	return c
+}
+
+func (c *CgroupResource) WithValidator(validator ResourceValidator) SystemResource {
+	c.Validator = validator
+	return c
+}
+```
+
+Registry of CgroupResource is defined as following. It is for generating and retrieving the various resource types of cgroups.
+
+```go
+func init() {
+	DefaultRegistry.Add(false, knownCgroupResources...)
+	DefaultRegistry.Add(true, knownCgroupV2Resources...)
+}
+
+var DefaultRegistry = NewCgroupResourceRegistry()
+
+type CgroupResourceRegistry interface {
+	Add(isV2 bool, s ...SystemResource)
+	Get(t ResourceType) (SystemResource, bool)
+	GetV2(t ResourceType) (SystemResource, bool)
+}
+
+var DefaultFactory = NewCgroupResourceFactory()
+
+type CgroupResourceFactory interface {
+	NewCgroup(filename string, subfs string) SystemResource
+	NewCgroupV2(filename string) SystemResource
+}
+
+var (
+	CPUStat          = DefaultFactory.NewCgroup(CPUStatFileName, CgroupCPUDir)
+	CPUCFSQuota      = DefaultFactory.NewCgroup(CPUCFSQuotaName, CgroupCPUDir)
+	MemoryLimit      = DefaultFactory.NewCgroup(MemoryLimitFileName, CgroupMemDir)
+	MemoryMin        = DefaultFactory.NewCgroup(MemMinFileName, CgroupMemDir).WithSupported(false)
```

Comment:
where does the `false` in WithSupported comes from, for example the cpu.bvt

Reply:
The initialization is removed. This field should be set in the runtime.  e.g. When the`MemoryMin` is firstly used, check the kernel version and set the `MemoryMin.Supported`.